clients/src/main/java/org/apache/kafka/clients/ClientUtils.java

```diff
@@ -106,8 +106,9 @@ public static ChannelBuilder createChannelBuilder(AbstractConfig config, Time ti
                 clientSaslMechanism, time, true, logContext);
     }
 
-    static List<InetAddress> resolve(String host, ClientDnsLookup clientDnsLookup) throws UnknownHostException {
-        InetAddress[] addresses = InetAddress.getAllByName(host);
+    static List<InetAddress> resolve(String host, ClientDnsLookup clientDnsLookup,
+                                     HostResolver hostResolver) throws UnknownHostException {
```

Comment:
This method and its usages look a bit weird to me now. It resolves an `host` based on `clientDnsLookup` and a `hostResolver` which resolves an `host` at its turn. We also have to pass the `HostResolver` wherever it is called now. I wonder if we should avoid this level of indirection here and directly encapsulate the entire logic in the `HostResolver`.

`HostResolver` would look like this:
```
public interface HostResolver {
    List<InetAddress> resolve(String host, ClientDnsLookup clientDnsLookup) throws UnknownHostException;
}
```

In `NodeConnectionState#currentAddress`, we could replace `addresses = ClientUtils.resolve(host, clientDnsLookup, hostResolver)` by `addresses = hostResolver.resolve(host, clientDnsLookup)`.

The `DefaultHostResolver` would do what `ClientUtils.resolve` does today and we can mock it in tests like you did.

Is this something that you have considered?

Reply:
@dajac I started making this change, and I don't think I like requiring the `HostResolver` to understand the `client.dns.lookup` config. In the same way that `MockTime` is only used to mock Java built-in methods, I think the `HostResolver` should only be responsible for the DNS resolution (either real or mocked), and that `ClientUtils` should be responsible for understanding the Kafka application behavior. That way, tests that mock DNS still test the actual `ClientDnsLookup` behavior, rather than relying on both the default resolver and any mocked resolvers handle it correctly. It also means that if we ever add an option for `ClientDnsLookup` or change its behavior, we don't need to remember to update all `HostResolver` implementations.

Reply:
Thanks @bob-barrett. I do agree with your point. Let's keep it as you suggested.